bus: support multiple dispatch in one transaction

bergquist opened the pull request:
this makes it possible to run multiple DispatchCtx
in one transaction. The TransactionManager will
start/end the transaction and pass the dbsession
in the context.Context variable

bergquist committed:
bus: support multiple dispatch in one transaction

this makes it possible to run multiple DispatchCtx
in one transaction. The TransactionManager will
start/end the transaction and pass the dbsession
in the context.Context variable

DanCech commented on pkg/services/sqlstore/stats.go:
```diff
@@ -133,6 +135,22 @@ func GetAdminStats(query *m.GetAdminStatsQuery) error {
+		return err
+	})
+}
+
```
Is there a reason for adding this besides a demonstration? It seems like if we want to have `ensureAdminUser` use a transaction it would need to also wrap the dispatch of the create user command.

bergquist replied:
Nope. Should have mentioned that in the PR

DanCech replied:
Probably still a good idea to wrap the create user command in the transaction to make it a good demonstration

bergquist committed:
replace begin/end with wrapper function

DanCech commented on pkg/services/sqlstore/sqlstore.go:
```diff
@@ -94,37 +94,49 @@ func (ss *SqlStore) Init() error {
 	if err != nil {
 		sess.Rollback()
 		return err
+	} else if err = sess.Commit(); err != nil {
```
else here is not needed after return above

DanCech commented on pkg/services/sqlstore/sqlstore.go:
```diff
@@ -94,37 +94,49 @@ func (ss *SqlStore) Init() error {
-	if err != nil {
-		return ctx, err
-	}
+func (stm *SQLTransactionManager) Wrapp(ctx context.Context, fn func(ctx context.Context) error) error {
```
any reason for the doubled `p`?

bergquist replied:
bad grammer

bergquist committed:
check if admin exists or create one in one transaction

bergquist committed:
fixes typo in code

torkelo committed:
refactoring: transaction manager PR #12203

torkelo committed:
refactoring: renamed AddCtxHandler to AddHandlerCtx PR #12203

torkelo commented:
This looked great but had a few issues that I had a hard time to pinpoint so started experimenting a bit making some changes in the code. 

So maybe not best PR review practice, if any of the changes below is making this worse please revert. 

* Changed interface in bus package from TransactionWrapper to TransactionManager to make naming more consistent (since it is named TransactionManager everwhere else) 
* Changed interface function name from Wrap to InTransaction
* Made SqlStore implement it directly instead of having a separate type (this is to make it more accessible/discoverable internally in SqlStore package and elsewhere)
* To make code dealing with transactions less spread out moved transaction stuff from shared.go and sqlstore.go to it's own file (transactions.go), and renamed shared.go to session.go as all that remained is DBSession code. 
* Tried in vain to get rid of the duplicate inTransactionWithRetyCtx methods but the different signatures and one being instance method and one static makes it a bit tricky so abandoned that for now. Now the two duplicated functions are in the same file close to each other at least. 
* Lastly in a seperate commit I did a unrelated change (to this PR), noticed an inconsistency in the naming of the AddCtxHandler PR, usually overloaded methods are given a postfix like DispatchCtx. So to align the naming conventions here I renamed AddCtxHandler to AddHandlerCtx

DanCech commented:
Looks pretty good to me, the refactor makes sense.

I'd argue for providing `inTransactionWithCtx` to avoid having to specify the `retry` parameter every time we want to use a transaction with context, also when `inTransactionWithRetryCtx` tries to retry should it be calling itself and passing the context through, rather than calling `inTransactionWithRetry` (which creates a new context)?

The big thing that still worries me is the need to have 2 functions for eg `CreateUser` and `CreateUserCtx`. Could we instead have `Dispatch` be able to call functions that are registered with context by creating a background context, since all that happens now is that `inTransactionWithRetry` creates `context.Background()` and uses that?

torkelo commented:
> I'd argue for providing inTransactionWithCtx 

Should not be a need for this as handlers can use the InTransaction Sqlstore instance function that already takes ctx  (the handler method needs to changed into instance method but that is the plan). 
https://github.com/bergquist/grafana/blob/8521e8f7b77ce0bdcce6192883ed6073c379ee50/pkg/services/sqlstore/transactions.go#L12

> rather than calling inTransactionWithRetry (which creates a new context)? 

Oh, yes! I also noticied that the retry keeps creating nested contexts. The problem is that for each retry the session is closed and and a new created. So a new context might be needed unless you can update a context value. 

> The big thing that still worries me is the need to have 2 functions for eg CreateUser and CreateUserCtx.  

Yea, that is not good. Did not notice that.  Also not liking the Ctx postfix of these handler methods, feels unnecessary unless we need 2 variants of the same handler which I cannot see a need for.

marefr commented:
This is definitely needed :+1: however would be nice if we could separate queries and commands. If command - a transaction will always be started. This approach is much more explicit in my mind and something to consider for the future.

bergquist commented:
> The big thing that still worries me is the need to have 2 functions for eg CreateUser and CreateUserCtx

Once this PR is accepted we should start to refactor all handler to accept `context.Context` as their first param. All handler should accept `context.Context` but can choose to ignore it. Otherwise, only some selected handlers can be used in `InTransaction`. 

We should only have 
```go
bus.Dispatch(ctx, msg)
```

> rather than calling inTransactionWithRetry (which creates a new context)? 

This is needed as Torkel mentioned. Some comments would make it more clear. 

> This is definitely needed +1 however would be nice if we could separate queries and commands. If command - a transaction will always be started. This approach is much more explicit in my mind and something to consider for the future.

That's an interesting idea. Kinda out of scope for this PR. Considering how much discussion this PR sparked I'll write up an issue for how we can improve the `bus`

bergquist committed:
bus: DispatchCtx can now invoke any handler

bergquist commented on pkg/api/dashboard.go:
```diff
@@ -103,7 +103,7 @@ func GetDashboard(c *m.ReqContext) Response {
 
 	isDashboardProvisioned := &m.IsDashboardProvisionedQuery{DashboardId: dash.Id}
-	err = bus.Dispatch(isDashboardProvisioned)
+	err = bus.DispatchCtx(c.Req.Context(), isDashboardProvisioned)
```
This is just for showing that `DispatchCtx` now can invoke both classic and ctx handlers.

torkelo replied:
Whats the need for the old one in that case? Why not have single Dispatch function?

bergquist replied:
@torkelo This makes it easier to transition one handler at the time.

bergquist committed:
bus: dont mix ctx/classic handlers

bergquist committed:
bus: dont start transaction when creating session

bergquist committed:
bus: Dispatch now passes empty ctx if handler require it

bergquist committed:
removes unused code

bergquist committed:
bus: noop should still execute fn

bergquist committed:
sql: adds tests for InTransaction

bergquist commented:
Merging this once 5.2-beta2 is released.

DanCech commented on pkg/services/sqlstore/user.go:
```diff
@@ -79,77 +81,87 @@ func getOrgIdForNewUser(cmd *m.CreateUserCommand, sess *DBSession) (int64, error
+
+func CreateUser(cmd *m.CreateUserCommand) error {
+	return inTransaction(func(sess *DBSession) error {
+		return internalCreateUser(sess, cmd)
```
We can revert most of this now and just register `CreateUserCtx` since `Dispatch` will take care of providing the context.  Some of the `InTransaction` variants could also be removed I think.

bergquist committed:
transactions: start sessions and transactions at the same place

this make it possible for handler to use `withSession` when
transactions is not nedded and `inTransactionCtx` if its needed
without knowing who owns the session/transaction

bergquist committed:
merge create user handlers

DanCech commented on pkg/services/sqlstore/user.go:
```diff
@@ -150,7 +152,7 @@ func CreateUser(cmd *m.CreateUserCommand) error {
 
 		return nil
-	})
+	}, 0)
```
This change is looking much cleaner now 🙂 I do still wonder whether we should provide `inTransactionCtx` that can then call `inTransactionWithRetryCtx(..., 0)` (in the same way `inTransaction` calls `inTransactionWithRetry`) to avoid having to add the `, 0` all over the place.

bergquist replied:
Yeap. I added that function in the `alert_reminder` branch to make things cleaner. might as well add it here since they might not be merged at the same time.

bergquist committed:
adds inTransactionCtx that calls inTransactionWithRetryCtx

DanCech reviewed (approved):
Looking good 😎

torkelo commented on pkg/bus/bus.go:
```diff
@@ -167,6 +205,20 @@ func Publish(msg Msg) error {
+// The caller can then pass a function with multiple DispatchCtx calls that
+// all will be executed in the same transaction. InTransaction will rollback if the
+// callback returns an error.
+func InTransaction(ctx context.Context, fn func(ctx context.Context) error) error {
```
do we need this? Where do we need a global access to this function? (just dont want to introduce new global stuff)

torkelo commented on pkg/services/sqlstore/transactions.go:
```diff
@@ -0,0 +1,106 @@
+	return inTransactionWithRetry(callback, 0)
+}
+
+func inTransactionCtx(ctx context.Context, callback dbTransactionFunc) error {
```
Do we need this?  can't we use the instance method Sqlstore.InTransaction defined above (don't want to add new global funcs if we can avoid it)

torkelo replied:
I see that it is used in CreateUser, this function could easily be changed to instance function (on Sqlstore) but then init function where handlers are registered also needs to be an instance func. No biggie, but we can leave that out of this PR :)

bergquist replied:
Migration to use the `Sqlstore.InTransaction` is the correct solution. But I feel this can be an ok intermediate step for migrating handlers without also having to start using the bus. To make the transitions as smooth as possible.
